{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cb95cae1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "862962b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jaspe\\AppData\\Local\\Temp\\ipykernel_70964\\4126571102.py:1: DtypeWarning: Columns (0,38,40) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\"clinvar_conflicting.csv\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNDISDBINCL</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>...</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>BLOSUM62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1168180</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>0.10020</td>\n",
       "      <td>0.1066</td>\n",
       "      <td>MedGen:CN169374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>...</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.053</td>\n",
       "      <td>-0.208682</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1470752</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1843891,OMIM:607454,Orphanet:ORPHA9877...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spinocerebellar_ataxia_21|not_provided</td>\n",
       "      <td>...</td>\n",
       "      <td>deleterious_low_confidence</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.000</td>\n",
       "      <td>6.517838</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1737942</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Human_Phenotype_Ontology:HP:0000486,MedGen:C00...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Strabismus|Nystagmus|Hypothyroidism|Intellectu...</td>\n",
       "      <td>...</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.100</td>\n",
       "      <td>6.061752</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome|not_provided</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.500</td>\n",
       "      <td>3.114491</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.700</td>\n",
       "      <td>4.766224</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS REF ALT  AF_ESP  AF_EXAC  AF_TGP  \\\n",
       "0     1  1168180   G   C  0.0771  0.10020  0.1066   \n",
       "1     1  1470752   G   A  0.0000  0.00000  0.0000   \n",
       "2     1  1737942   A   G  0.0000  0.00001  0.0000   \n",
       "3     1  2160305   G   A  0.0000  0.00000  0.0000   \n",
       "4     1  2160305   G   T  0.0000  0.00000  0.0000   \n",
       "\n",
       "                                            CLNDISDB CLNDISDBINCL  \\\n",
       "0                                    MedGen:CN169374          NaN   \n",
       "1  MedGen:C1843891,OMIM:607454,Orphanet:ORPHA9877...          NaN   \n",
       "2  Human_Phenotype_Ontology:HP:0000486,MedGen:C00...          NaN   \n",
       "3  MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002...          NaN   \n",
       "4     MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002          NaN   \n",
       "\n",
       "                                               CLNDN  ...  \\\n",
       "0                                      not_specified  ...   \n",
       "1             Spinocerebellar_ataxia_21|not_provided  ...   \n",
       "2  Strabismus|Nystagmus|Hypothyroidism|Intellectu...  ...   \n",
       "3          Shprintzen-Goldberg_syndrome|not_provided  ...   \n",
       "4                       Shprintzen-Goldberg_syndrome  ...   \n",
       "\n",
       "                         SIFT           PolyPhen MOTIF_NAME MOTIF_POS  \\\n",
       "0                   tolerated             benign        NaN       NaN   \n",
       "1  deleterious_low_confidence             benign        NaN       NaN   \n",
       "2                 deleterious  probably_damaging        NaN       NaN   \n",
       "3                         NaN                NaN        NaN       NaN   \n",
       "4                         NaN                NaN        NaN       NaN   \n",
       "\n",
       "  HIGH_INF_POS MOTIF_SCORE_CHANGE  LoFtool  CADD_PHRED  CADD_RAW BLOSUM62  \n",
       "0          NaN                NaN      NaN       1.053 -0.208682      2.0  \n",
       "1          NaN                NaN      NaN      31.000  6.517838     -3.0  \n",
       "2          NaN                NaN      NaN      28.100  6.061752     -1.0  \n",
       "3          NaN                NaN      NaN      22.500  3.114491      NaN  \n",
       "4          NaN                NaN      NaN      24.700  4.766224     -3.0  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"clinvar_conflicting.csv\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dcb9729b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rows: 65188\n",
      "Columns: 46\n",
      "CHROM                  object\n",
      "POS                     int64\n",
      "REF                    object\n",
      "ALT                    object\n",
      "AF_ESP                float64\n",
      "AF_EXAC               float64\n",
      "AF_TGP                float64\n",
      "CLNDISDB               object\n",
      "CLNDISDBINCL           object\n",
      "CLNDN                  object\n",
      "CLNDNINCL              object\n",
      "CLNHGVS                object\n",
      "CLNSIGINCL             object\n",
      "CLNVC                  object\n",
      "CLNVI                  object\n",
      "MC                     object\n",
      "ORIGIN                  int64\n",
      "SSR                   float64\n",
      "CLASS                   int64\n",
      "Allele                 object\n",
      "Consequence            object\n",
      "IMPACT                 object\n",
      "SYMBOL                 object\n",
      "Feature_type           object\n",
      "Feature                object\n",
      "BIOTYPE                object\n",
      "EXON                   object\n",
      "INTRON                 object\n",
      "cDNA_position          object\n",
      "CDS_position           object\n",
      "Protein_position       object\n",
      "Amino_acids            object\n",
      "Codons                 object\n",
      "DISTANCE              float64\n",
      "STRAND                float64\n",
      "BAM_EDIT               object\n",
      "SIFT                   object\n",
      "PolyPhen               object\n",
      "MOTIF_NAME             object\n",
      "MOTIF_POS             float64\n",
      "HIGH_INF_POS           object\n",
      "MOTIF_SCORE_CHANGE    float64\n",
      "LoFtool               float64\n",
      "CADD_PHRED            float64\n",
      "CADD_RAW              float64\n",
      "BLOSUM62              float64\n",
      "dtype: object\n",
      "Class distribution:\n",
      "CLASS\n",
      "0    48754\n",
      "1    16434\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(\"Rows:\", df.shape[0])\n",
    "print(\"Columns:\", df.shape[1])\n",
    "print(df.dtypes)\n",
    "\n",
    "print(\"Class distribution:\")\n",
    "print(df['CLASS'].value_counts())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4eff6840",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='CLASS', ylabel='count'>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAk0AAAGwCAYAAAC0HlECAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjMsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvZiW1igAAAAlwSFlzAAAPYQAAD2EBqD+naQAAJjZJREFUeJzt3QuQleV5B/AHkJsgEO5SQDSmIopQuUkTrRgixsvUURs0Vom3jJQYEQtKaiBaM6Q6VjAiaKzBdLQi6RiLF9TBS6piUJQKKIwmtJIqFy+AErlv5/2m58wuoL4ieHaX32/mzNnv+579zrtnZtk/7+00qKqqqgoAAD5Vw0+/DABAIjQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACDDfjlFfLbt27fH22+/HQcccEA0aNCg0s0BADKk7So//PDD6NKlSzRs+Ol9SULTHpICU7du3SrdDABgN6xYsSK6du36qTVC0x6SephKb3qrVq0q3RwAIMP69euLTo/S3/FPIzTtIaUhuRSYhCYAqFtyptaYCA4AkEFoAgCo7aHpJz/5SdEdVv3Rs2fP8vWNGzfGqFGjol27dtGyZcs488wzY9WqVTXu8dZbb8Upp5wS+++/f3Ts2DHGjh0bW7durVHz9NNPx9FHHx1NmzaNQw89NGbMmLFTW6ZOnRo9evSIZs2axaBBg2L+/Pl78ScHAOqaivc0HXHEEfHOO++UH88++2z52hVXXBGzZ8+OWbNmxTPPPFOsUDvjjDPK17dt21YEps2bN8fzzz8fd999dxGIJkyYUK5Zvnx5UTNkyJBYuHBhjB49Oi6++OJ47LHHyjUzZ86MMWPGxMSJE+Pll1+OPn36xLBhw2L16tVf4jsBANRqVRU0ceLEqj59+uzy2tq1a6saN25cNWvWrPK5119/vSo1ed68ecXxI488UtWwYcOqlStXlmumTZtW1apVq6pNmzYVx+PGjas64ogjatx7+PDhVcOGDSsfDxw4sGrUqFHl423btlV16dKlatKkSdk/y7p164q2pWcAoG74PH+/K97T9MYbbxQbSh1yyCFx7rnnFsNtyYIFC2LLli0xdOjQcm0auuvevXvMmzevOE7PvXv3jk6dOpVrUg9RWj64ZMmSck31e5RqSvdIvVTptarXpM2t0nGpZlc2bdpUvE71BwBQf1U0NKW5Q2k4bc6cOTFt2rRiKO3YY48tduZcuXJlNGnSJNq0aVPje1JASteS9Fw9MJWul659Wk0KOR9//HG8++67xTDfrmpK99iVSZMmRevWrcsPG1sCQP1W0X2avv3tb5e/Puqoo4oQddBBB8X9998fzZs3j9ps/PjxxTyoHTfHAgDqp4oPz1WXepX+/M//PN58883o3LlzMXS2du3aGjVp9Vy6lqTnHVfTlY4/qyZtQJmCWfv27aNRo0a7rCndY1fSSrzSRpY2tASA+q9WhaaPPvoofv/738eBBx4Y/fr1i8aNG8fcuXPL15ctW1bMeRo8eHBxnJ4XLVpUY5XbE088UQSYXr16lWuq36NUU7pHGgJMr1W9Jn34bjou1QAAVHT13JVXXln19NNPVy1fvrzqueeeqxo6dGhV+/btq1avXl1cv/TSS6u6d+9e9eSTT1a99NJLVYMHDy4eJVu3bq068sgjq0488cSqhQsXVs2ZM6eqQ4cOVePHjy/X/OEPf6jaf//9q8aOHVusvps6dWpVo0aNitqS++67r6pp06ZVM2bMqHrttdeqvv/971e1adOmxqq8z2L1HADUPZ/n73dF5zT98Y9/jHPOOSfee++96NChQ3zjG9+IF154ofg6ufnmm4uVbGlTy7RaLa16u+2228rfn4bVHnrooRg5cmTRK9SiRYsYMWJEXHfddeWagw8+OB5++OFiz6cpU6YUn2B85513FvcqGT58eKxZs6bY3ylN/u7bt28xOX3HyeEAwL6rQUpOlW5EfZAmgqdVdOvWrTO/CQDq4d/vWjWnCQCgthKaAAAyVHROE59fv7G/qnQToNZZcOP5lW4CsA/Q0wQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgCoS6HpZz/7WTRo0CBGjx5dPrdx48YYNWpUtGvXLlq2bBlnnnlmrFq1qsb3vfXWW3HKKafE/vvvHx07doyxY8fG1q1ba9Q8/fTTcfTRR0fTpk3j0EMPjRkzZuz0+lOnTo0ePXpEs2bNYtCgQTF//vy9+NMCAHVNrQhNL774Ytx+++1x1FFH1Th/xRVXxOzZs2PWrFnxzDPPxNtvvx1nnHFG+fq2bduKwLR58+Z4/vnn4+677y4C0YQJE8o1y5cvL2qGDBkSCxcuLELZxRdfHI899li5ZubMmTFmzJiYOHFivPzyy9GnT58YNmxYrF69+kt6BwCA2q5BVVVVVSUb8NFHHxW9QLfddltcf/310bdv35g8eXKsW7cuOnToEPfee2+cddZZRe3SpUvj8MMPj3nz5sUxxxwTjz76aJx66qlFmOrUqVNRM3369LjqqqtizZo10aRJk+Lrhx9+OBYvXlx+zbPPPjvWrl0bc+bMKY5Tz9KAAQPi1ltvLY63b98e3bp1i8suuyyuvvrqrJ9j/fr10bp166LdrVq1ir2l39hf7bV7Q1214MbzK90EoI76PH+/K97TlIbfUk/Q0KFDa5xfsGBBbNmypcb5nj17Rvfu3YvQlKTn3r17lwNTknqI0huwZMmScs2O9041pXukXqr0WtVrGjZsWByXanZl06ZNxetUfwAA9dd+lXzx++67rxgOS8NzO1q5cmXRU9SmTZsa51NAStdKNdUDU+l66dqn1aSQ8/HHH8cHH3xQDPPtqib1bH2SSZMmxbXXXvu5f2YAoG6qWE/TihUr4vLLL4977rmnmHxd14wfP77oyis90s8DANRfFQtNaUgsTbRO85n222+/4pEme99yyy3F16mnJw2dpblH1aXVc507dy6+Ts87rqYrHX9WTRq3bN68ebRv3z4aNWq0y5rSPXYlrcRL96j+AADqr4qFpm9+85uxaNGiYkVb6dG/f/8499xzy183btw45s6dW/6eZcuWFVsMDB48uDhOz+ke1Ve5PfHEE0WA6dWrV7mm+j1KNaV7pCHAfv361ahJE8HTcakGAKBic5oOOOCAOPLII2uca9GiRbEnU+n8RRddVGwF0LZt2yIIpdVsKciklXPJiSeeWISj8847L2644YZi/tI111xTTC5PPUHJpZdeWqyKGzduXFx44YXx5JNPxv3331+sqCtJrzFixIgiqA0cOLBYvbdhw4a44IILvtT3BACovSo6Efyz3HzzzcVKtrSpZVqtlla9pa0JStKw2kMPPRQjR44swlQKXSn8XHfddeWagw8+uAhIac+nKVOmRNeuXePOO+8s7lUyfPjwYouCtL9TCl5p24O0HcGOk8MBgH1Xxfdpqi/s0wSVY58mYJ/YpwkAoC4QmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJAKC2h6Zp06bFUUcdFa1atSoegwcPjkcffbR8fePGjTFq1Kho165dtGzZMs4888xYtWpVjXu89dZbccopp8T+++8fHTt2jLFjx8bWrVtr1Dz99NNx9NFHR9OmTePQQw+NGTNm7NSWqVOnRo8ePaJZs2YxaNCgmD9//l78yQGAuqaioalr167xs5/9LBYsWBAvvfRSnHDCCfHXf/3XsWTJkuL6FVdcEbNnz45Zs2bFM888E2+//XacccYZ5e/ftm1bEZg2b94czz//fNx9991FIJowYUK5Zvny5UXNkCFDYuHChTF69Oi4+OKL47HHHivXzJw5M8aMGRMTJ06Ml19+Ofr06RPDhg2L1atXf8nvCABQWzWoqqqqilqkbdu2ceONN8ZZZ50VHTp0iHvvvbf4Olm6dGkcfvjhMW/evDjmmGOKXqlTTz21CFOdOnUqaqZPnx5XXXVVrFmzJpo0aVJ8/fDDD8fixYvLr3H22WfH2rVrY86cOcVx6lkaMGBA3HrrrcXx9u3bo1u3bnHZZZfF1Vdfvct2btq0qXiUrF+/vviedevWFb1me0u/sb/aa/eGumrBjedXuglAHZX+frdu3Trr73etmdOUeo3uu+++2LBhQzFMl3qftmzZEkOHDi3X9OzZM7p3716EpiQ99+7duxyYktRDlN6AUm9Vqql+j1JN6R6plyq9VvWahg0bFselml2ZNGlS8SaXHikwAQD1V8VD06JFi4r5Smm+0aWXXhoPPPBA9OrVK1auXFn0FLVp06ZGfQpI6VqSnqsHptL10rVPq0nB6uOPP4533323CGy7qindY1fGjx9fpNLSY8WKFV/wnQAAarP9Kt2Aww47rJhrlILHr3/96xgxYkQxf6m2SyEvPQCAfUPFQ1PqTUor2pJ+/frFiy++GFOmTInhw4cXQ2dp7lH13qa0eq5z587F1+l5x1VupdV11Wt2XHGXjtO4ZfPmzaNRo0bFY1c1pXsAAFR8eG5HaRJ2mmCdAlTjxo1j7ty55WvLli0rthhIc56S9JyG96qvcnviiSeKQJSG+Eo11e9RqindI4W29FrVa1Ib0nGpBgCgoj1NaV7Qt7/97WJy94cfflislEt7KqXtANLk6osuuqjYCiCtqEtBKK1mS0EmrZxLTjzxxCIcnXfeeXHDDTcUc5CuueaaYm+n0tBZmieVVsWNGzcuLrzwwnjyySfj/vvvL1bUlaTXSMOC/fv3j4EDB8bkyZOLCekXXHBBxd4bAKB2qWhoSj1E559/frzzzjtFSEobXabA9K1vfau4fvPNNxcr2dKmlqn3Ka16u+2228rfn4bVHnrooRg5cmQRplq0aFGEn+uuu65cc/DBBxcBKe35lIb90t5Qd955Z3GvkjQUmLYoSPs7peDVt2/fYjuCHSeHAwD7rlq3T9O+sM/DF2GfJtiZfZqAfWqfJgCA2kxoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwDA3gpNJ5xwQvGZcLva6yBdAwCob3YrNKWPOkkfprujjRs3xn/+53/uiXYBANTdj1F59dVXy1+/9tprxUeOlGzbtq346JE/+7M/27MtBACoa6EpfSZbgwYNiseuhuGaN28eP//5z/dk+wAA6l5oWr58eaSPqjvkkENi/vz50aFDh/K1Jk2aRMeOHYsP0QUA2KdD00EHHVQ8b9++fW+1BwCg7oem6t5444146qmnYvXq1TuFqAkTJuyJtgEA1O3Q9Itf/CJGjhwZ7du3j86dOxdznErS10ITAFDf7FZouv766+OnP/1pXHXVVXu+RQAA9WWfpg8++CD+5m/+Zs+3BgCgPoWmFJgef/zxPd8aAID6NDx36KGHxo9//ON44YUXonfv3tG4ceMa13/4wx/uqfYBANTd0HTHHXdEy5Yt45lnnike1aWJ4EITAFDf7FZoSptcAgDsS3ZrThMAwL5mt3qaLrzwwk+9ftddd+1uewAA6k9oSlsOVLdly5ZYvHhxrF27dpcf5AsAsE+GpgceeGCnc+mjVNIu4V/96lf3RLsAAOrnnKaGDRvGmDFj4uabb95TtwQAqJ8TwX//+9/H1q1b9+QtAQDq7vBc6lGqrqqqKt555514+OGHY8SIEXuqbQAAdTs0vfLKKzsNzXXo0CFuuummz1xZBwCwz4Smp556as+3BACgvoWmkjVr1sSyZcuKrw877LCitwkAoD7arYngGzZsKIbhDjzwwDjuuOOKR5cuXeKiiy6KP/3pT3u+lQAAdTE0pYng6YN6Z8+eXWxomR4PPvhgce7KK6/c860EAKiLw3P//u//Hr/+9a/j+OOPL587+eSTo3nz5vGd73wnpk2btifbCABQN3ua0hBcp06ddjrfsWNHw3MAQL20W6Fp8ODBMXHixNi4cWP53McffxzXXnttcQ0AoL7ZreG5yZMnx0knnRRdu3aNPn36FOf+67/+K5o2bRqPP/74nm4jAEDdDE29e/eON954I+65555YunRpce6cc86Jc889t5jXBABQ3+xWaJo0aVIxp+mSSy6pcf6uu+4q9m666qqr9lT7AADq7pym22+/PXr27LnT+SOOOCKmT5++J9oFAFD3Q9PKlSuLjS13lHYETx/cCwBQ3+xWaOrWrVs899xzO51P59LO4AAA9c1uzWlKc5lGjx4dW7ZsiRNOOKE4N3fu3Bg3bpwdwQGAemm3QtPYsWPjvffei7/7u7+LzZs3F+eaNWtWTAAfP378nm4jAEDdDE0NGjSIf/qnf4of//jH8frrrxfbDHzta18r9mkCAKiPdis0lbRs2TIGDBiw51oDAFCfJoIDAOxrhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBANT20DRp0qQYMGBAHHDAAdGxY8c4/fTTY9myZTVqNm7cGKNGjYp27dpFy5Yt48wzz4xVq1bVqHnrrbfilFNOif3337+4z9ixY2Pr1q01ap5++uk4+uijo2nTpnHooYfGjBkzdmrP1KlTo0ePHtGsWbMYNGhQzJ8/fy/95ABAXVPR0PTMM88UgeiFF16IJ554IrZs2RInnnhibNiwoVxzxRVXxOzZs2PWrFlF/dtvvx1nnHFG+fq2bduKwLR58+Z4/vnn4+677y4C0YQJE8o1y5cvL2qGDBkSCxcujNGjR8fFF18cjz32WLlm5syZMWbMmJg4cWK8/PLL0adPnxg2bFisXr36S3xHAIDaqkFVVVVV1BJr1qwpeopSODruuONi3bp10aFDh7j33nvjrLPOKmqWLl0ahx9+eMybNy+OOeaYePTRR+PUU08twlSnTp2KmunTp8dVV11V3K9JkybF1w8//HAsXry4/Fpnn312rF27NubMmVMcp56l1Ot16623Fsfbt2+Pbt26xWWXXRZXX331Tm3dtGlT8ShZv359UZ/a3KpVq732HvUb+6u9dm+oqxbceH6lmwDUUenvd+vWrbP+fteqOU2pwUnbtm2L5wULFhS9T0OHDi3X9OzZM7p3716EpiQ99+7duxyYktRDlN6EJUuWlGuq36NUU7pH6qVKr1W9pmHDhsVxqWZXQ4vpTS49UmACAOqvWhOaUs9OGjb7+te/HkceeWRxbuXKlUVPUZs2bWrUpoCUrpVqqgem0vXStU+rScHq448/jnfffbcY5ttVTekeOxo/fnwR8kqPFStWfOH3AACovfaLWiLNbUrDZ88++2zUBWlCeXoAAPuGWtHT9IMf/CAeeuiheOqpp6Jr167l8507dy6GztLco+rS6rl0rVSz42q60vFn1aSxy+bNm0f79u2jUaNGu6wp3QMA2LdVNDSlOegpMD3wwAPx5JNPxsEHH1zjer9+/aJx48Yxd+7c8rm0JUHaYmDw4MHFcXpetGhRjVVuaSVeCkS9evUq11S/R6mmdI80BJheq3pNGi5Mx6UaAGDftl+lh+TSyrgHH3yw2KupNH8oTaxOPUDp+aKLLiq2AkiTw1MQSqvZUpBJK+eStEVBCkfnnXde3HDDDcU9rrnmmuLepeGzSy+9tFgVN27cuLjwwguLgHb//fcXK+pK0muMGDEi+vfvHwMHDozJkycXWx9ccMEFFXp3AIDapKKhadq0acXz8ccfX+P8L3/5y/je975XfH3zzTcXK9nSppZpiX9a9XbbbbeVa9OwWhraGzlyZBGmWrRoUYSf6667rlyTerBSQEp7Pk2ZMqUYArzzzjuLe5UMHz682KIg7e+Uglffvn2L7Qh2nBwOAOybatU+TfvKPg9fhH2aYGf2aQL2uX2aAABqK6EJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACCD0AQAkEFoAgDIIDQBAGQQmgAAMghNAAAZhCYAgAxCEwBABqEJACDDfjlFAOx9/cb+qtJNgFpnwY3nR22hpwkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBABQ20PTb3/72zjttNOiS5cu0aBBg/jNb35T43pVVVVMmDAhDjzwwGjevHkMHTo03njjjRo177//fpx77rnRqlWraNOmTVx00UXx0Ucf1ah59dVX49hjj41mzZpFt27d4oYbbtipLbNmzYqePXsWNb17945HHnlkL/3UAEBdVNHQtGHDhujTp09MnTp1l9dTuLnlllti+vTp8bvf/S5atGgRw4YNi40bN5ZrUmBasmRJPPHEE/HQQw8VQez73/9++fr69evjxBNPjIMOOigWLFgQN954Y/zkJz+JO+64o1zz/PPPxznnnFMErldeeSVOP/304rF48eK9/A4AAHVFg6rUnVMLpJ6mBx54oAgrSWpW6oG68sor4+///u+Lc+vWrYtOnTrFjBkz4uyzz47XX389evXqFS+++GL079+/qJkzZ06cfPLJ8cc//rH4/mnTpsU//MM/xMqVK6NJkyZFzdVXX130ai1durQ4Hj58eBHgUugqOeaYY6Jv375FYMuRwlnr1q2LNqZer72l39hf7bV7Q1214Mbzoz7w+w1f/u/35/n7XWvnNC1fvrwIOmlIriT9UIMGDYp58+YVx+k5DcmVAlOS6hs2bFj0TJVqjjvuuHJgSlJv1bJly+KDDz4o11R/nVJN6XV2ZdOmTcUbXf0BANRftTY0pcCUpJ6l6tJx6Vp67tixY43r++23X7Rt27ZGza7uUf01PqmmdH1XJk2aVIS40iPNlQIA6q9aG5pqu/HjxxddeaXHihUrKt0kAGBfDE2dO3cunletWlXjfDouXUvPq1evrnF969atxYq66jW7ukf11/ikmtL1XWnatGkx9ln9AQDUX7U2NB188MFFaJk7d275XJo3lOYqDR48uDhOz2vXri1WxZU8+eSTsX379mLuU6kmrajbsmVLuSattDvssMPiK1/5Srmm+uuUakqvAwBQ0dCU9lNauHBh8ShN/k5fv/XWW8VqutGjR8f1118f//Ef/xGLFi2K888/v1gRV1phd/jhh8dJJ50Ul1xyScyfPz+ee+65+MEPflCsrEt1yXe/+91iEnjaTiBtTTBz5syYMmVKjBkzptyOyy+/vFh1d9NNNxUr6tKWBC+99FJxLwCAZL9Kvg0pmAwZMqR8XAoyI0aMKLYVGDduXLEVQNp3KfUofeMb3yjCTdqAsuSee+4pws03v/nNYtXcmWeeWeztVJImaT/++OMxatSo6NevX7Rv377YMLP6Xk5/+Zd/Gffee29cc8018aMf/Si+9rWvFVsSHHnkkV/aewEA1G61Zp+mus4+TVA59mmC+muBfZoAAOoWoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBACQQWgCAMggNAEAZBCadjB16tTo0aNHNGvWLAYNGhTz58+vdJMAgFpAaKpm5syZMWbMmJg4cWK8/PLL0adPnxg2bFisXr260k0DACpMaKrmn//5n+OSSy6JCy64IHr16hXTp0+P/fffP+66665KNw0AqLD9Kt2A2mLz5s2xYMGCGD9+fPlcw4YNY+jQoTFv3ryd6jdt2lQ8StatW1c8r1+/fq+2c9umj/fq/aEu2tu/d18Wv9/w5f9+l+5fVVX1mbVC0/979913Y9u2bdGpU6ca59Px0qVLd6qfNGlSXHvttTud79at215tJ7Cz1j+/tNJNAOr47/eHH34YrVu3/tQaoWk3pR6pNP+pZPv27fH+++9Hu3btokGDBhVtG3tf+p9JCsgrVqyIVq1aVbo5wB7k93vfUlVVVQSmLl26fGat0PT/2rdvH40aNYpVq1bVOJ+OO3fuvFN906ZNi0d1bdq02evtpHZJ/6D6RxXqJ7/f+47Wn9HDVGIi+P9r0qRJ9OvXL+bOnVuj9ygdDx48uKJtAwAqT09TNWm4bcSIEdG/f/8YOHBgTJ48OTZs2FCspgMA9m1CUzXDhw+PNWvWxIQJE2LlypXRt2/fmDNnzk6TwyENzab9vHYcogXqPr/ffJIGVTlr7AAA9nHmNAEAZBCaAAAyCE0AABmEJgCADEIT7IapU6dGjx49olmzZjFo0KCYP39+pZsEfEG//e1v47TTTit2hk6f7PCb3/ym0k2ilhGa4HOaOXNmsadXWpL88ssvR58+fWLYsGGxevXqSjcN+ALSvnzp9zn9pwh2xZYD8DmlnqUBAwbErbfeWt45Pn1O1WWXXRZXX311pZsH7AGpp+mBBx6I008/vdJNoRbR0wSfw+bNm2PBggUxdOjQ8rmGDRsWx/Pmzato2wDYu4Qm+Bzefffd2LZt2067xKfjtIs8APWX0AQAkEFogs+hffv20ahRo1i1alWN8+m4c+fOFWsXAHuf0ASfQ5MmTaJfv34xd+7c8rk0ETwdDx48uKJtA2Dv2m8v3x/qnbTdwIgRI6J///4xcODAmDx5crFU+YILLqh004Av4KOPPoo333yzfLx8+fJYuHBhtG3bNrp3717RtlE72HIAdkPabuDGG28sJn/37ds3brnllmIrAqDuevrpp2PIkCE7nU//SZoxY0ZF2kTtIjQBAGQwpwkAIIPQBACQQWgCAMggNAEAZBCaAAAyCE0AABmEJgCADEITAEAGoQkAIIPQBNRL6SNuLrvssjjkkEOiadOm0a1btzjttNPKH7bco0eP4nMDP8u//du/RaNGjWLUqFG7vP6LX/wi+vTpEy1btow2bdrEX/zFX8SkSZPK1//0pz/F+PHj46tf/Wo0a9YsOnToEH/1V38VDz744B78aYEvgw/sBeqd//7v/46vf/3rRYhJnxHYu3fv2LJlSzz22GNF+Fm6dGn2vf7lX/4lxo0bF7fffnvcdNNNRfApueuuu2L06NHFZw+mILRp06Z49dVXY/HixeWaSy+9NH73u9/Fz3/+8+jVq1e899578fzzzxfPQN3is+eAeufkk08uwsuyZcuiRYsWNa6tXbu2CFOppykFnvT4JOlT7o844oh45513YtiwYfHDH/4wvvvd75avn3766fGVr3wlfvnLX37iPdJrTZkypfjQV6BuMzwH1Cvvv/9+zJkzp+hR2jEwlUJMrhSGTjnllGjdunX87d/+bdHrVF3nzp3jhRdeiP/5n//5xHukmkceeSQ+/PDDz/mTALWN0ATUK2+++WakDvSePXt+ofts3749ZsyYUYSl5Oyzz45nn3226H0qmThxYrnX6rDDDovvfe97cf/99xffW3LHHXcUw3Ht2rWLAQMGxBVXXBHPPffcF2obUBlCE1Cv7KkZB0888URs2LChGOpL2rdvH9/61reKeUwlBx54YMybNy8WLVoUl19+eWzdurUYhjvppJPKwem4446LP/zhD8UE9LPOOiuWLFkSxx57bPzjP/7jHmkn8OUxpwmod8NzKeD89Kc/LVatfZLPmtP0ne98J2bNmlWsnCtJQahr167FRPOGDXf9f87UG5VC0ZNPPhlDhgzZZc31118f1113XXz00UfRpEmTz/0zApWhpwmoV9q2bVtM2p46dWrRU7SjNBH8s6SVbWlLgPvuuy8WLlxYfrzyyivxwQcfxOOPP/6J35tWyCW7eu3qNalXauPGjdk/F1B5thwA6p0UmNKWAwMHDix6dI466qgipKQht2nTpsXrr79e1P3v//5vEYaqO+igg+Jf//VfizlIqbepQYMGNa6n4bo0ITwNwY0cOTK6dOkSJ5xwQtEDlVbZpV6ktBfT4MGDi/rjjz8+zjnnnOjfv39xz9deey1+9KMfFb1QrVq1+hLfFeCLMjwH1EspwKQhuoceeqj4OgWZfv36FROxU5BJw3O7WvWWAtMNN9xQDLGl8LWjNNH7vPPOKwLXM888U8xxSj1QqXcqDQumsJQmiKe9oZK00eXs2bOL7Q/SRpcpZJ166qkxYcKEIkQBdYfQBACQwZwmAIAMQhMAQAahCQAgg9AEAJBBaAIAyCA0AQBkEJoAADIITQAAGYQmAIAMQhMAQAahCQAgPtv/Ad8Vz+7NLWLYAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.countplot(x='CLASS', data=df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2ae6cae5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature columns: ['CHROM', 'POS', 'REF', 'ALT', 'AF_ESP', 'AF_EXAC', 'AF_TGP', 'CLNDISDB', 'CLNDISDBINCL', 'CLNDN', 'CLNDNINCL', 'CLNHGVS', 'CLNSIGINCL', 'CLNVC', 'CLNVI', 'MC', 'ORIGIN', 'SSR', 'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Feature_type', 'Feature', 'BIOTYPE', 'EXON', 'INTRON', 'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids', 'Codons', 'DISTANCE', 'STRAND', 'BAM_EDIT', 'SIFT', 'PolyPhen', 'MOTIF_NAME', 'MOTIF_POS', 'HIGH_INF_POS', 'MOTIF_SCORE_CHANGE', 'LoFtool', 'CADD_PHRED', 'CADD_RAW', 'BLOSUM62']\n"
     ]
    }
   ],
   "source": [
    "features = df.drop(columns=['CLASS'])\n",
    "labels = df['CLASS']\n",
    "print(\"feature columns:\", features.columns.tolist())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0410f0a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "overall means:\n",
      "POS                   7.757594e+07\n",
      "AF_ESP                1.451052e-02\n",
      "AF_EXAC               1.449218e-02\n",
      "AF_TGP                1.526350e-02\n",
      "ORIGIN                1.342486e+00\n",
      "SSR                   2.269231e+00\n",
      "DISTANCE              8.257315e+02\n",
      "STRAND               -6.659097e-03\n",
      "MOTIF_POS             1.000000e+00\n",
      "MOTIF_SCORE_CHANGE   -8.000000e-02\n",
      "LoFtool               3.450584e-01\n",
      "CADD_PHRED            1.568562e+01\n",
      "CADD_RAW              2.554131e+00\n",
      "BLOSUM62             -4.022584e-01\n",
      "dtype: float64\n",
      "overall stddevs:\n",
      "POS                   5.974051e+07\n",
      "AF_ESP                5.779541e-02\n",
      "AF_EXAC               5.954210e-02\n",
      "AF_TGP                5.952741e-02\n",
      "ORIGIN                5.688772e+00\n",
      "SSR                   4.190777e+00\n",
      "DISTANCE              1.069363e+03\n",
      "STRAND                9.999855e-01\n",
      "MOTIF_POS             0.000000e+00\n",
      "MOTIF_SCORE_CHANGE    2.404163e-02\n",
      "LoFtool               3.612384e-01\n",
      "CADD_PHRED            1.083635e+01\n",
      "CADD_RAW              2.961553e+00\n",
      "BLOSUM62              1.872684e+00\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(\"overall means:\")\n",
    "print(features.select_dtypes(include='number').mean())\n",
    "\n",
    "print(\"overall stddevs:\")\n",
    "print(features.select_dtypes(include='number').std())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "94ec1635",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train: (39112, 45) Test: (13038, 45) Val: (13038, 45)\n",
      "Train class dist:\n",
      " CLASS\n",
      "0    0.747903\n",
      "1    0.252097\n",
      "Name: proportion, dtype: float64\n",
      "Test class dist:\n",
      " CLASS\n",
      "0    0.747891\n",
      "1    0.252109\n",
      "Name: proportion, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "X_temp, X_val, y_temp, y_val = train_test_split(features, labels, test_size=0.20, random_state=0, stratify=labels) \n",
    "#validation\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(X_temp, y_temp, test_size=0.25, random_state=0, stratify=y_temp)\n",
    "#splits train and test\n",
    "\n",
    "print(\"Train:\", X_train.shape, \"Test:\", X_test.shape, \"Val:\", X_val.shape)\n",
    "#prints row/columns\n",
    "print(\"Train class dist:\\n\", y_train.value_counts(normalize=True))\n",
    "print(\"Test class dist:\\n\", y_test.value_counts(normalize=True))\n",
    "#gives percentages\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "dc9ebede",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   Data Type  Missing Values  Missing Percentage  \\\n",
      "CHROM                 object               0            0.000000   \n",
      "POS                    int64               0            0.000000   \n",
      "REF                   object               0            0.000000   \n",
      "ALT                   object               0            0.000000   \n",
      "AF_ESP               float64               0            0.000000   \n",
      "AF_EXAC              float64               0            0.000000   \n",
      "AF_TGP               float64               0            0.000000   \n",
      "CLNDISDB              object               0            0.000000   \n",
      "CLNDN                 object               0            0.000000   \n",
      "CLNHGVS               object               0            0.000000   \n",
      "CLNVC                 object               0            0.000000   \n",
      "IMPACT                object               0            0.000000   \n",
      "Consequence           object               0            0.000000   \n",
      "Allele                object               0            0.000000   \n",
      "CLASS                  int64               0            0.000000   \n",
      "ORIGIN                 int64               0            0.000000   \n",
      "Feature               object              14            0.021476   \n",
      "Feature_type          object              14            0.021476   \n",
      "STRAND               float64              14            0.021476   \n",
      "BIOTYPE               object              16            0.024544   \n",
      "SYMBOL                object              16            0.024544   \n",
      "MC                    object             846            1.297785   \n",
      "CADD_PHRED           float64            1092            1.675155   \n",
      "CADD_RAW             float64            1092            1.675155   \n",
      "LoFtool              float64            4213            6.462846   \n",
      "cDNA_position         object            8884           13.628275   \n",
      "EXON                  object            8893           13.642081   \n",
      "Protein_position      object            9955           15.271216   \n",
      "CDS_position          object            9955           15.271216   \n",
      "Amino_acids           object           10004           15.346383   \n",
      "Codons                object           10004           15.346383   \n",
      "BAM_EDIT              object           33219           50.958765   \n",
      "CLNVI                 object           37529           57.570412   \n",
      "BLOSUM62             float64           39595           60.739707   \n",
      "SIFT                  object           40352           61.900963   \n",
      "PolyPhen              object           40392           61.962324   \n",
      "INTRON                object           56385           86.495981   \n",
      "CLNDISDBINCL          object           65021           99.743818   \n",
      "CLNDNINCL             object           65021           99.743818   \n",
      "CLNSIGINCL            object           65021           99.743818   \n",
      "SSR                  float64           65058           99.800577   \n",
      "DISTANCE             float64           65080           99.834325   \n",
      "MOTIF_POS            float64           65186           99.996932   \n",
      "MOTIF_NAME            object           65186           99.996932   \n",
      "HIGH_INF_POS          object           65186           99.996932   \n",
      "MOTIF_SCORE_CHANGE   float64           65186           99.996932   \n",
      "\n",
      "                    Unique Values  \n",
      "CHROM                          38  \n",
      "POS                         63115  \n",
      "REF                           866  \n",
      "ALT                           458  \n",
      "AF_ESP                       2842  \n",
      "AF_EXAC                      6667  \n",
      "AF_TGP                       2087  \n",
      "CLNDISDB                     9234  \n",
      "CLNDN                        9260  \n",
      "CLNHGVS                     65188  \n",
      "CLNVC                           7  \n",
      "IMPACT                          4  \n",
      "Consequence                    48  \n",
      "Allele                        374  \n",
      "CLASS                           2  \n",
      "ORIGIN                         31  \n",
      "Feature                      2369  \n",
      "Feature_type                    2  \n",
      "STRAND                          2  \n",
      "BIOTYPE                         2  \n",
      "SYMBOL                       2328  \n",
      "MC                             90  \n",
      "CADD_PHRED                   9324  \n",
      "CADD_RAW                    63803  \n",
      "LoFtool                      1195  \n",
      "cDNA_position               13970  \n",
      "EXON                         3264  \n",
      "Protein_position             7339  \n",
      "CDS_position                13663  \n",
      "Amino_acids                  1262  \n",
      "Codons                       2220  \n",
      "BAM_EDIT                        2  \n",
      "CLNVI                       27654  \n",
      "BLOSUM62                        6  \n",
      "SIFT                            4  \n",
      "PolyPhen                        4  \n",
      "INTRON                       1929  \n",
      "CLNDISDBINCL                   93  \n",
      "CLNDNINCL                     101  \n",
      "CLNSIGINCL                    137  \n",
      "SSR                             2  \n",
      "DISTANCE                       96  \n",
      "MOTIF_POS                       1  \n",
      "MOTIF_NAME                      2  \n",
      "HIGH_INF_POS                    1  \n",
      "MOTIF_SCORE_CHANGE              2  \n"
     ]
    }
   ],
   "source": [
    "feature_summary = pd.DataFrame({\n",
    "    'Data Type': df.dtypes,\n",
    "    'Missing Values': df.isnull().sum(), #returns true or false based on missing or not\n",
    "    'Missing Percentage': df.isnull().sum() / len(df) * 100, #percentage of misisng values, takes mising values and divides by total * 100\n",
    "    'Unique Values': df.nunique()\n",
    "    \n",
    "})\n",
    "\n",
    "print(feature_summary.sort_values(by='Missing Percentage')) #sorts by missing percent\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9570042a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dropped columns: ['CLNDISDBINCL', 'CLNDNINCL', 'CLNSIGINCL', 'SSR', 'INTRON', 'DISTANCE', 'SIFT', 'PolyPhen', 'MOTIF_NAME', 'MOTIF_POS', 'HIGH_INF_POS', 'MOTIF_SCORE_CHANGE', 'BLOSUM62']\n",
      "shape: (65188, 32)\n"
     ]
    }
   ],
   "source": [
    "temp = feature_summary['Missing Percentage'] #creates list of missing percents\n",
    "columnsmissing = feature_summary[temp > 60].index #takes the feature summary table, and compares temp with 60, and gives true or false, before making a table of the missing vals\n",
    "features = features.drop(columns=columnsmissing)\n",
    "\n",
    "print(\"dropped columns:\", columnsmissing.tolist())\n",
    "print(\"shape:\", features.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b738dd0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropped columns: ['REF', 'ALT', 'CLNDISDB', 'CLNDISDBINCL', 'CLNDN', 'CLNDNINCL', 'CLNHGVS', 'CLNSIGINCL', 'CLNVI', 'MC', 'Allele', 'SYMBOL', 'Feature', 'EXON', 'INTRON', 'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids', 'Codons']\n",
      "One-hot encoded columns: ['CHROM', 'CLNVC', 'Consequence', 'IMPACT', 'Feature_type', 'BIOTYPE', 'BAM_EDIT', 'SIFT', 'PolyPhen', 'MOTIF_NAME', 'HIGH_INF_POS']\n",
      "\n",
      "Final Check\n",
      "\n",
      "Total missing values left: 0\n",
      "Data shape:  (65188, 114)\n"
     ]
    }
   ],
   "source": [
    "# PREPROCESSING!\n",
    "\n",
    "\n",
    "df = pd.read_csv('clinvar_conflicting.csv', low_memory=False)\n",
    "\n",
    "features = df.drop(columns=['CLASS']) # drops class variable  to prep model to predict later\n",
    "\n",
    "def preprocessing(df, uniquelimit=50):\n",
    "    # preprocessing\n",
    "    df_processed = df.copy()\n",
    "    categorical_cols = df_processed.select_dtypes(include=['object']).columns #finds all object (string/text) datatypes\n",
    "\n",
    "    cols_to_drop = []\n",
    "    for col in categorical_cols: \n",
    "        if df_processed[col].nunique() > uniquelimit:    \n",
    "            cols_to_drop.append(col) #adds all to-drop columns to cols_to_drop\n",
    "\n",
    "\n",
    "    df_processed = df_processed.drop(columns=cols_to_drop) #drops\n",
    "    print(f\"Dropped columns: {cols_to_drop}\")\n",
    "\n",
    "    cols_to_encode = []\n",
    "    for col in categorical_cols:\n",
    "        if col not in cols_to_drop:\n",
    "            cols_to_encode.append(col) #columns to onehot encode\n",
    "\n",
    "\n",
    "    df_processed = pd.get_dummies(df_processed, columns=cols_to_encode, dummy_na=False) #one hot encoding, transforms df_processed into numerical, and dummy_na makes it not create a row for missing values, but to keep it.\n",
    "    print(f\"One-hot encoded columns: {cols_to_encode}\")\n",
    "    return df_processed\n",
    "\n",
    "def fillmissingval(df):\n",
    "    #fills missing data w/ median\n",
    "    df_imputed = df.copy()\n",
    "\n",
    "    for col in df_imputed.columns:\n",
    "        if df_imputed[col].isnull().any():\n",
    "            median_val = df_imputed[col].median()\n",
    "            df_imputed[col] = df_imputed[col].fillna(median_val) #fills missing values in dataset w/ median\n",
    "\n",
    "    return df_imputed\n",
    "\n",
    "features_encoded = preprocessing(features) \n",
    "features_final = fillmissingval(features_encoded)\n",
    "\n",
    "print(\"\\nFinal Check:\")\n",
    "print(\"\\nTotal missing values left:\", features_final.isnull().sum().sum()) #sum of missing values in each column to check for missed missing values\n",
    "print(\"Data shape: \", features_final.shape) #(rows, columns)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7553e085",
   "metadata": {},
   "source": [
    "## CLINVAR DATA CLEANING REPORT\n",
    "\n",
    "### Introduction\n",
    "\n",
    "This report explains the data preprocessing steps for clinvar_conflicting_csv. The goal is to transform the data into a clean, numerical format, in order to train it with a binary classification model.\n",
    "\n",
    "### Feature Analysis\n",
    "\n",
    "The original dataset of clinvar_conflicting_csv contains 45 features and 1 target variable (CLASS). \n",
    "The quantitative data in the dataset included the following columns:\n",
    "POS, AF_ESP, AF_EXAC, AF_TGP, ORIGIN, SSR, DISTANCE, STRAND, MOTIF_POS, MOTIF_SCORE_CHANGE, LoFtool, CADD_PHRED, CADD_RAW, BLOSUM62\n",
    "\n",
    "The categorical features that are suitable for one-hot encoding:\n",
    "CHROM, CLNVC, Consequence, IMPACT, Feature_type, BIOTYPE, BAM_EDIT.\n",
    "\n",
    "The remaining columns won't be used, as they all contain a high amount of unique text entries, making them inefficient for one-hot encoding.\n",
    "\n",
    "### Missing Value Analysis\n",
    "\n",
    "Many of the features contains missing values. In order to handle this, we have to first identify which columns have missing values, and what percentage of missing data makes a column missing. I did this by checking if >60% of the data in the columns are missing. The list of missing data includes:\n",
    "MOTIF_POS, MOTIF_NAME, SIFT, PolyPhen, BLOSUM62, INTRON\n",
    "\n",
    "### Preprocessing\n",
    "\n",
    "After dropping the high missing data columns, the next step was to drop highly unique columns (those with >50 unique values). If we didn't do this, it would lead to the curse of dimensionality, as when one-hot encoding, it creates a column for every unique category. Therefore, if a cetgory had thousands of unique values, it would create thousands of new columns, leading to overfitting and inefficiency. Dropping the highly unique columns removes unnecessary noise, making the program more efficient.\n",
    "\n",
    "### One-hot encoding\n",
    "\n",
    "One-hot encoding is the process of convering all categorical / qualitative variables into quantitative variables. This is done by following the steps:\n",
    "1. Find the unique categories\n",
    "2. Create new columns, with a binary 0 or 1 for each category\n",
    "3. Places 1s and 0s coresponding to the original data, in the correct row / column\n",
    "4. Drops the original text column, as now everything is represented numerically\n",
    "\n",
    "\n",
    "One-hot encoding prevents the model from incorrectly learning relations, as if we just assigned each unique value to a number, it would think higher numbers are more valuable than lower numbers.\n",
    "\n",
    "### Filling in values\n",
    "\n",
    "After the dropping and dencoding steps, with the numerical data, the last step is to fill in all the blank missing data. I did this by finding the median of the column, and filling all missing data with this median. This works pretty well, as it gives an average of the dataset without being impacted by skews. However, one thing to note is that if there is an overwhelming about of missing data, by replacing all the missing data with the same value, it could skew the data a bit. \n",
    "\n",
    "\n",
    "\n",
    "Comments are in the code block right above that explains the preprocessing steps.\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
